Sources/Bookmarks/BookmarksDatabaseCleaner.swift

```diff
@@ -0,0 +1,88 @@
+//
+//  BookmarkDatabaseCleaner.swift
+//  DuckDuckGo
+//
+//  Copyright © 2023 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import Combine
+import Common
+import CoreData
+import Persistence
+
+public struct BookmarksCleanupError: Error {
+    public let coreDataError: Error
+}
+
+public final class BookmarkDatabaseCleaner {
+
+    public init(bookmarkDatabase: CoreDataDatabase, errorEvents: EventMapping<BookmarksCleanupError>?) {
+        self.database = bookmarkDatabase
+        self.errorEvents = errorEvents
+
+        cleanupCancellable = triggerSubject
+            .receive(on: workQueue)
+            .sink { [weak self] _ in
+                self?.removeBookmarksPendingDeletion()
+            }
+    }
+
+    public func scheduleRegularCleaning() {
+        cancelCleaningSchedule()
+        scheduleCleanupCancellable = Timer.publish(every: Const.cleanupInterval, on: .main, in: .default)
+            .sink { [weak self] _ in
+                self?.triggerSubject.send()
+            }
+    }
+
+    public func cancelCleaningSchedule() {
+        scheduleCleanupCancellable?.cancel()
+    }
+
+    public func cleanUpDatabaseNow() {
+        triggerSubject.send()
+    }
+
+    private func removeBookmarksPendingDeletion() {
+        let context = database.makeContext(concurrencyType: .privateQueueConcurrencyType)
+
+        context.performAndWait {
+            let bookmarksPendingDeletion = BookmarkUtils.fetchBookmarksPendingDeletion(context)
+
+            for bookmark in bookmarksPendingDeletion {
+                context.delete(bookmark)
+            }
+
+            do {
+                try context.save()
+            } catch {
+                errorEvents?.fire(.init(coreDataError: error))
```

Comment:
No something to think about: we don't really have merge strategies set up, so we need to consider two things:
- Save failures here, caused by change coming from other context - unlikely as we are recreating the context but possible due to time window we have between fetch and save.
- Save failure in all other Bookmarks contexts - also should not happen often as we are merging but is possible.

I was initially thinking this will need to be brought up once we start saving sync context, but since we've introduced that async cleanup, we most likely shall be doing TD for that problem now.

Reply:
Following our Zoom call, I’ve added a mechanism that retries cleanup on context merge errors.